simulator/mips/_instr.h

```diff
@@ -0,0 +1,85 @@
+enum instr_size_and_shifts {
+    FUNCT_SIZE    = 6,
+    SHAMT_SIZE    = 5,
+    RD_SIZE       = 5,
+    RT_SIZE       = 5,
+    RS_SIZE       = 5,
+    OPCODE_SIZE   = 6,
+    ASI_IMM_SIZE  = 16,
+    ASJ_IMM_SIZE  = 26,
+
+    FUNCT_SHIFT   = 0,
+    SHAMT_SHIFT   = FUNCT_SHIFT + FUNCT_SIZE,
+    RD_SHIFT      = SHAMT_SHIFT + SHAMT_SIZE,
+    RT_SHIFT      = RD_SHIFT    + RD_SIZE,
+    RS_SHIFT      = RT_SHIFT    + RT_SIZE,
+    OPCODE_SHIFT  = RS_SHIFT    + RS_SIZE,
+    ASI_IMM_SHIFT = 0,
+    ASJ_IMM_SHIFT = 0,
+};
+
+                              //32bits          01234567012345670123456701234567
+static const constexpr uint32 FUNCT_MASK    = 0b00000000000000000000000000111111;
```

Comment:
In modern C++ you may separate literal with quote mark, like that: `0b00000011'11100000`. That would make code more readable.